### DISCUSSION TITLE
Automation Delete row trigger with query and relations

### DISCUSSION THREAD

**Discussion by harmkroog** [Budibase/budibase#10461] [Help] [answered]
Good day,
Ive got  a problem and dont undestand it.

Ive got to tables.
Table A with customers
An Table B with adresses.
The table B is associated in relationship with table a.
The DB format is budibase internal so ive got an field ADRESSNREL in table kunden  with 
following formula return $("adressen")[0]._id.
because the internal DB, relatioship are not listen in design. So ive created an second field that conatis the value from relaatioship field.
And the same in the table adresses, for the field kunden wich is an relatioship ive cretaed an field KUNDENREL.

So now to my problem

Ive created an automation. If in table kunden an customer is deleted then a query runs with filter in table adresses
_ID equals {{ trigger.row.ADRESSENREL }}

What shoud it do:
it should return the rows with the adresses from the deleted customer.
Then after the query an delete action will run over the adresses.
but how can i delete only adresses really from the deleted customer.

But i think the problem is, in the moment if the customer is deleted, the reference in adress is gone.
So that this get an error. How can i prevent this.
I thought about, to delete all adresses without reference. 
In the Budibase data view, it called Error while ex...
I define an filter in the query as value .
I used the starts with compare function 
Error while ex

This works only in test.
In real life the trigger row delete, nothing work to. 
In the automation history, nothing done:
My system is an docker image from budibase. 
Must i create anything more for automation in docker? 
or why it ist not trigger if an row in table kudnen is deleted
I hope you can understand me.
Thank you, very much!
Best regards Bonkersdeluxe

**Comment by ConorWebb96** [comment]
Hey @harmkroog,

Could you send us some screenshots of your automaton. So we can better understand what you are doing. If possible an export of your app would be nice as well.  This should be something which works in Budibase, from the details you've provided I'm not quite sure where you are going wrong.

On a side note make sure your app is **published**, as this can cause issues with automations as they only run on published apps.

Thanks in advance.

**Comment by harmkroog** [comment]
Hi, thank you for reply
What i would.
i try to explain.

In the First screenshot you see the customers.
![Auswahl_062](https://user-images.githubusercontent.com/130768693/235796018-225964ab-2ba2-4a4f-ac78-ac9ede2fe96b.png)
For every customer you can assign adresses over the button adresses.
With typ Invoice and delivery adresses. But the typ is unnessary here.
![Auswahl_063](https://user-images.githubusercontent.com/130768693/235796340-8209dfc2-3be3-48c8-8546-447f3d7ff964.png)

Now the Problem i get.
If i delete an customer, i will delete all Adresses assigned to customer to.
The Database structure

Table Kunden
The Field Adressenrel is an pointer to Adressen. Because its an Budibaseinteral DB, relations not work on filter and gui elements in the designer.

![Auswahl_064](https://user-images.githubusercontent.com/130768693/235796898-9f0eed19-a815-4da6-9a00-0f6c3669db65.png)
So use an workaround. The Javascript in the Field Adressenrel
![Auswahl_065](https://user-images.githubusercontent.com/130768693/235797036-ce09ed6a-870e-48fe-ae4b-f39e4da9874d.png)
![Auswahl_066](https://user-images.githubusercontent.com/130768693/235797113-98fc4985-2ed7-4364-a1f8-bdac85dba44f.png)

The same at the adressen Table
![Auswahl_067](https://user-images.githubusercontent.com/130768693/235797764-542d1fc9-4ea6-4d18-a14b-bc0de2595c57.png)
![Auswahl_068](https://user-images.githubusercontent.com/130768693/235797831-dd37806b-8d7e-487b-ab82-badc74b453a3.png)
![Auswahl_069](https://user-images.githubusercontent.com/130768693/235797875-5e43c2fd-8a37-4f63-9adb-f5a304b2dbd3.png)

Now to the Automation, thank you for the notice, that automations not work in preview Mode. I dont know this before.
I think thats why the delete trigger not triggered.

But how i define the query.
because if the customer is deleted, the reference to the adress is lost.
Contains the trigger.row the complete old dataset stored?
Then i could query the adressid with adressrel from the deleted dataset?

![Auswahl_070](https://user-images.githubusercontent.com/130768693/235798209-e39e2484-384c-46ee-b1f0-cd519ec0f62f.png)
The define in query
![Auswahl_071](https://user-images.githubusercontent.com/130768693/235798354-d7b47274-99dc-4be1-b903-193372436d80.png)
The test run
![Auswahl_072](https://user-images.githubusercontent.com/130768693/235798427-b35d7663-bfb7-42d2-bd9e-ef7adb550cf2.png)
And the log from testrun:
`{
  "tableId": "ta_12c02974d0364e949e8d2f5b13c974ff",
  "filters": {
    "string": {},
    "fuzzy": {},
    "range": {},
    "equal": {
      "1:_id": "{{ trigger.row.ADRESSENREL }}"
    },
    "notEqual": {},
    "empty": {},
    "notEmpty": {},
    "contains": {},
    "notContains": {},
    "oneOf": {},
    "containsAny": {},
    "allOr": true
  },
  "filters-def": [
    {
      "id": "hEzj3UHn1",
      "field": "1:_id",
      "operator": "equal",
      "value": "{{ trigger.row.ADRESSENREL }}",
      "valueType": "Binding",
      "type": "string",
      "noValue": false
    },
    {
      "operator": "allOr"
    }
  ],
  "sortColumn": "ADRESSENREL",
  "sortOrder": "ascending",
  "onEmptyFilter": "none"
}`

I would export the application. I have deleted all rows with attachment and added test data.
but in export the attachment files would be in. Ive deleted all rows.
why the attachment still exist. How can i delete the attachments? Ive deleted all rows.
Why he is saving it?
If i post the application, it should contains only fictive data.
Please how can i remove the the attachment files?
![Auswahl_073](https://user-images.githubusercontent.com/130768693/235801851-d217e478-366b-4c05-a163-3c29ab978406.png)
![Auswahl_074](https://user-images.githubusercontent.com/130768693/235801940-84016a9e-2492-481f-bc93-fa16dbdb6fea.png)

Then i can upload the exported Application

Thank you very much!
Best regards Bonkersdeluxe

**Reply by ConorWebb96** [reply]
Hey @harmkroog,

Could you also add a screenshot of the **Output** of the error you are getting for the below screenshot?
![235798427-b35d7663-bfb7-42d2-bd9e-ef7adb550cf2](https://user-images.githubusercontent.com/126772285/235862206-5021803c-2f42-44d6-a038-d09a24a5b052.png)

**Sidenote**, I believe the main issue here is that related items are also being deleted whenever you are deleting an address. The Budibase tables are cascading meaning that related data tends to be deleted as well. 

I think in this case you may be better off using a different DB for example SQL. As it simply would be better for a complicated use-case such as yours. Filtering on relationships also works in SQL, meaning you would have an easier time here as well. Overall, I feel like you may be trying to fight the default functionality that comes with the Budibase DB. Below I've linked a good video on which DB may be best suited for what you want to do.

https://docs.budibase.com/docs/data#getting-started-with-data

I hope all of this makes sense

P.S. Thank you for all the information you've provided so far.

**Comment by harmkroog** [comment]
Hi, Thank you!
Here is the Output.
{
  "rows": [],
  "success": false
}
ill will watch the video.
Thank you!
Best regards Bonkersdeluxe

**Comment by andz-bb** [comment, marked as answer]
hey @harmkroog 

if I'm understanding the problem correctly you want an automation that deletes all related addresses when a customer is deleted.

![delete related columns demo](https://user-images.githubusercontent.com/110921612/235925683-b3d5f373-9a23-4cba-9a9a-42cf84fc1ca6.gif)

Here's how I achieved it:

I created a static formula column `ADRESSENREL` on my Kunden table but instead of returning the first _id I am returning all of them. Here is the JS I am using:

```js
return $("Adressen") ? $("Adressen").map(x => x._id) : ""
```

In a case where I have two addresses for a customer, the following is returned: `["ro_ta_b56ca96daf08496a8c3f0b1eef049ad0_5061edf7f0f6422a8a61e8f1dcc35a5f","ro_ta_b56ca96daf08496a8c3f0b1eef049ad0_8935ed6f7684464980b760d01a5d9b55"]`

When deleting the Kunden row that `ADRESSENREL` value is available on the automation trigger. You can then use it as an input for a looped delete row action.

<img width="504" alt="image" src="https://user-images.githubusercontent.com/110921612/235928508-334899ad-19f5-4706-a354-0a6afa719064.png">

Each Adresse _id for the deleted customer is provided to the delete row action so everything is cleared when the customer is deleted. I hope this is what you were looking for! If you have any questions, let me know 😊

**Comment by harmkroog** [comment]
It works, meeeegaaaaa.
Thank you very much!!!!!
I never get myself on this solution. Meeeegaaaaa!
Thank you very much!
Last Question you recommend mysql.
Is the handling the same, eg. create fields in budibase or must i use phpmyadmin or something else?
Best regards Bonkersdeluxe!
